core/src/main/java/org/testcontainers/images/RemoteDockerImage.java

```diff
@@ -0,0 +1,85 @@
+package org.testcontainers.images;
+
+import com.github.dockerjava.api.DockerClient;
+import com.github.dockerjava.api.model.Image;
+import com.github.dockerjava.core.command.PullImageResultCallback;
+import lombok.NonNull;
+import org.slf4j.profiler.Profiler;
+import org.testcontainers.DockerClientFactory;
+import org.testcontainers.containers.ContainerFetchException;
+import org.testcontainers.utility.DockerLoggerFactory;
+import org.testcontainers.utility.LazyFuture;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class RemoteDockerImage extends LazyFuture<String> {
+
+    private static final Set<String> AVAILABLE_IMAGE_NAME_CACHE = new HashSet<>();
+
+    private final String dockerImageName;
+
+    public RemoteDockerImage(@NonNull String repository, String tag) {
+        this.dockerImageName = repository + ":" + tag;
+    }
+
+    @Override
+    protected final String resolve() {
+        Profiler profiler = new Profiler("Rule creation - prefetch image");
+        profiler.setLogger(DockerLoggerFactory.getLogger(dockerImageName));
+
+        try (DockerClient dockerClient = DockerClientFactory.instance().client()) {
+
+            profiler.start("Check local images");
+
+            int attempts = 0;
+            while (true) {
+                // Does our cache already know the image?
+                if (AVAILABLE_IMAGE_NAME_CACHE.contains(dockerImageName)) {
```

Comment:
You'll see from my (only substantial) change on the previous PR that I moved this if block to before the dockerClient creation. This seems to save ~100-200ms per invocation, as creating the client seems to be a bit time consuming. Of course, a pool in `DockerClientFactory` might be an improvement for further down the line too :)